starknet-rpc-test/src/lib.rs

```diff
@@ -128,6 +155,9 @@ impl MadaraClient {
     }
 
     pub async fn new(execution: ExecutionStrategy) -> Self {
+        // we keep the reference, otherwise the mutex unlocks immediately
+        let _mutex_guard = FREE_PORT_ATTRIBUTION_MUTEX.lock().await;
+
```

Comment:
Out of curiosity, why was the first impl not working?

Reply:
So previously, we acquired the lock during the process of finding the port and released the lock when we found a free port. We didn't wait till the current thread launched the madara instance and acquired this free port. So what would happen is

1. Thread 1 took the lock
2. Thread 1 found out port 1025 is open
3. Thread 1 released the lock
4. Thread 1 fired up the madara instance
5. Thread 2 started look for ports and found 1025 open as well (thread 1 is still starting the madara instance so it hasn't acquired port 1025 yet)
6. Thread 1 starts on port 1025. Thread 2 madara instance starts on some other port but it makes calls to 1025 only thinking that's the instance assigned to it.
7. If Thread 1 test cases gets over before Thread 2, it will close the Madara instance and Thread 2 will fail.

So the fix was just to release the lock once Madara is fired up and has acquired the port.

Reply:
Thx